test/ecto/repo/many_to_many_test.exs

```diff
@@ -426,6 +427,28 @@ defmodule Ecto.Repo.ManyToManyTest do
     assert_received {:delete_all, _}
   end
 
+  test "deleting assocs with join_where on update on_replace" do
+    sample = %MyAssoc{id: 10, x: "xyz"} |> Ecto.put_meta(state: :loaded)
+
+    # removing assoc with join_where
+    %MySchema{id: 3, assocs: [sample], where_assocs: [sample]}
+    |> Ecto.Changeset.change()
+    |> Ecto.Changeset.put_assoc(:where_assocs, [])
+    |> TestRepo.update!()
+
+    # assert 3 where conditions: parent join key, child join key and join_where
+    assert_received {:delete_all, %{wheres: [_, _, _]}}
```

Comment:
Maybe get the query and call `inspect(...)` on it and match using `=~` on part of the query representation?